tests/test_2025_12_10_mainnet_v3.py

```diff
@@ -0,0 +1,796 @@
+            try:
+                oracle_daemon_config.get('SLASHING_RESERVE_WE_LEFT_SHIFT')
+                assert False, "SLASHING_RESERVE_WE_LEFT_SHIFT should not exist before vote"
+            except Exception:
+                pass  # Expected to fail
+
+            if details["status"] == PROPOSAL_STATUS["submitted"]:
+                chain.sleep(timelock.getAfterSubmitDelay() + 1)
+                dual_governance.scheduleProposal(EXPECTED_DG_PROPOSAL_ID, {"from": stranger})
+
+            if timelock.getProposalDetails(EXPECTED_DG_PROPOSAL_ID)["status"] == PROPOSAL_STATUS["scheduled"]:
+                chain.sleep(timelock.getAfterScheduleDelay() + 1)
+
+                wait_for_target_time_to_satisfy_time_constrains()
+
+                dg_tx: TransactionReceipt = timelock.execute(EXPECTED_DG_PROPOSAL_ID, {"from": stranger})
+                display_dg_events(dg_tx)
+                dg_events = group_dg_events_from_receipt(
+                    dg_tx,
+                    timelock=EMERGENCY_PROTECTED_TIMELOCK,
+                    admin_executor=DUAL_GOVERNANCE_ADMIN_EXECUTOR,
+                )
+                assert count_vote_items_by_events(dg_tx, agent.address) == EXPECTED_DG_EVENTS_FROM_AGENT
+                assert len(dg_events) == EXPECTED_DG_EVENTS_COUNT
+
+                # === DG EXECUTION EVENTS VALIDATION ===
+
+                # 1.2. Call V3Template.startUpgrade
+                assert 'UpgradeStarted' in dg_events[1]
+                assert dg_events[1]['UpgradeStarted'][0]['_emitted_by'] == UPGRADE_TEMPLATE
+
+                # 1.3. Lido Locator upgrade events
+                validate_proxy_upgrade_event(dg_events[2], LIDO_LOCATOR_IMPL, emitted_by=lido_locator_proxy)
+
+                # 1.4. Grant Aragon APP_MANAGER_ROLE to the AGENT
+                assert 'SetPermission' in dg_events[3]
+                assert dg_events[3]['SetPermission'][0]['allowed'] is True
+                assert dg_events[3]['SetPermission'][0]['_emitted_by'] == ACL
+                assert dg_events[3]['SetPermission'][0]['entity'] == AGENT
+                assert dg_events[3]['SetPermission'][0]['role'] == app_manager_role.hex()
+
+                # 1.5. Set Lido implementation in Kernel
+                assert 'SetApp' in dg_events[4]
+                assert dg_events[4]['SetApp'][0]['appId'] == LIDO_APP_ID
+                assert dg_events[4]['SetApp'][0]['_emitted_by'] == ARAGON_KERNEL
+                assert dg_events[4]['SetApp'][0]['app'] == LIDO_IMPL
+
+                # 1.6. Revoke Aragon APP_MANAGER_ROLE from the AGENT
+                assert 'SetPermission' in dg_events[5]
+                assert dg_events[5]['SetPermission'][0]['allowed'] is False
+                assert dg_events[5]['SetPermission'][0]['_emitted_by'] == ACL
+                assert dg_events[5]['SetPermission'][0]['entity'] == AGENT
+                assert dg_events[5]['SetPermission'][0]['role'] == app_manager_role.hex()
```

Comment:
consider wrapping events checks into complete event validators functions within this file to check more event params:
- that only expected events are fired (and no unexpected are fired), for example